File: Native/src/main/java/com/example/mindflip_native/views/PracticeScreen.kt
package com.example.mindflip_native.views

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.ArrowBack
import androidx.compose.material.icons.filled.ArrowForward
import androidx.compose.material.icons.filled.Close
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.runtime.livedata.observeAsState
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.*
import androidx.navigation.NavController
import com.example.mindflip_native.models.Flashcard
import com.example.mindflip_native.viewModels.MainViewModel


@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun PracticeScreen(
    navController: NavController,
    viewModel: MainViewModel
) {
    val flashcards by viewModel.flashcards.observeAsState(emptyList())

    var currentIndex by remember { mutableIntStateOf(0) }
    var showAnswer by remember { mutableStateOf(false) }

    val total = flashcards.size
    val currentFlashcard: Flashcard? = flashcards.getOrNull(currentIndex)

    if (currentFlashcard == null) {
        Box(
            modifier = Modifier
                .fillMaxSize()
                .background(Color(0xFFF0F0F0)),
            contentAlignment = Alignment.Center
        ) {
            Text("No flashcards available for this category.")
        }
        return
    }

    Scaffold(
        containerColor = Color(0xFFF0F0F0),
        topBar = {
            Column {
                CenterAlignedTopAppBar(
                    title = { Text("Practice") },
                    navigationIcon = {
                        IconButton(onClick = { navController.popBackStack() }) {
                            Icon(Icons.Default.Close, contentDescription = "Back", tint = Color.Black)
                        }
                    },
                    colors = TopAppBarDefaults.centerAlignedTopAppBarColors(
                        containerColor = Color(0xFFF0F0F0)
                    )
                )
                Divider(color = Color.Black.copy(alpha = 0.07f), thickness = 1.dp)
            }
        },
        bottomBar = {
            // Navigation buttons
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(horizontal = 20.dp, vertical = 12.dp),
                horizontalArrangement = Arrangement.SpaceBetween
            ) {
                Button(
                    onClick = {
                        if (currentIndex > 0) {
                            currentIndex--
                            showAnswer = false
                        }
                    },
                    enabled = currentIndex > 0,
                    colors = ButtonDefaults.buttonColors(containerColor = Color(0xFF1280ED)),
                    modifier = Modifier.size(width = 170.dp, height = 50.dp),
                    shape = RoundedCornerShape(12.dp)
                ) {
                    Icon(Icons.Default.ArrowBack, contentDescription = "Previous", tint = Color.White)
                    Spacer(Modifier.width(8.dp))
                    Text("Previous", color = Color.White, fontSize = 16.sp)
                }

                Button(
                    onClick = {
                        if (currentIndex < total - 1) {
                            currentIndex++
                            showAnswer = false
                        }
                    },
                    enabled = currentIndex < total - 1,
                    colors = ButtonDefaults.buttonColors(containerColor = Color(0xFFFFA726)),
                    modifier = Modifier.size(width = 170.dp, height = 50.dp),
                    shape = RoundedCornerShape(12.dp)
                ) {
                    Text("Next", color = Color.White, fontSize = 16.sp)
                    Spacer(Modifier.width(8.dp))
                    Icon(Icons.Default.ArrowForward, contentDescription = "Next", tint = Color.White)
                }
            }
        }
    ) { paddingValues ->
        Column(
            modifier = Modifier
                .padding(paddingValues)
                .fillMaxSize()
                .padding(horizontal = 24.dp),
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.Top
        ) {
            Spacer(modifier = Modifier.height(40.dp)) // more space under top bar

            // Progress info
            Text(
                text = "${currentIndex + 1}/$total",
                fontWeight = FontWeight.Medium,
                color = Color.Gray,
                fontSize = 16.sp
            )

            Spacer(modifier = Modifier.height(20.dp))

            // Blue progress bar
            LinearProgressIndicator(
                progress = (currentIndex + 1f) / total.toFloat(),
                modifier = Modifier
                    .fillMaxWidth()
                    .height(8.dp)
                    .clip(RoundedCornerShape(4.dp)),
                color = Color(0xFF1280ED),
                trackColor = Color.LightGray
            )

            Spacer(modifier = Modifier.height(60.dp))

            Box(
                modifier = Modifier
                    .fillMaxWidth()
                    .fillMaxHeight(0.55f)
                    .background(Color.White, RoundedCornerShape(20.dp))
                    .clickable { showAnswer = !showAnswer }
                    .padding(20.dp),
                contentAlignment = Alignment.Center
            ) {
                Column(
                    modifier = Modifier.fillMaxSize(),
                    verticalArrangement = Arrangement.Top,
                    horizontalAlignment = Alignment.Start
                ) {
                    // Difficulty box
                    val (bgColor, textColor) = when (currentFlashcard.difficulty.name) {
                        "EASY" -> Pair(Color(0x334CAF50), Color(0xFF4CAF50))
                        "MEDIUM" -> Pair(Color(0x33FFA726), Color(0xFFFFA726))
                        "HARD" -> Pair(Color(0x33F44336), Color(0xFFF44336))
                        else -> Pair(Color.LightGray.copy(alpha = 0.3f), Color.Gray)
                    }

                    Box(
                        modifier = Modifier
                            .background(bgColor, RoundedCornerShape(8.dp))
                            .padding(horizontal = 12.dp, vertical = 6.dp)
                    ) {
                        Text(
                            text = currentFlashcard.difficulty.name,
                            color = textColor,
                            fontWeight = FontWeight.Bold
                        )
                    }

                    Spacer(modifier = Modifier.height(24.dp))

                    // Dynamic font size based on text length (smooth scaling)
                    val text = if (showAnswer) currentFlashcard.answer else currentFlashcard.question
                    val dynamicFontSize = remember(text) {
                        val baseSize = 45f
                        val minSize = 14f
                        val scale = (baseSize - (text.length * 0.25f)).coerceIn(minSize, baseSize)
                        scale.sp
                    }

                    Text(
                        text = text,
                        fontSize = dynamicFontSize,
                        fontWeight = FontWeight.Medium,
                        color = Color.Black,
                        textAlign = TextAlign.Start,
                        modifier = Modifier
                            .align(Alignment.Start)
                            .fillMaxWidth()
                    )
                }
            }

            Spacer(modifier = Modifier.height(36.dp))

            // Show/Hide Button
            Button(
                onClick = { showAnswer = !showAnswer },
                colors = ButtonDefaults.buttonColors(
                    containerColor = if (showAnswer) Color(0xFFFFA726) else Color(0xFF1280ED) // 🔸 orange when showing answer
                ),
                modifier = Modifier
                    .fillMaxWidth(0.75f)
                    .height(70.dp),
                shape = RoundedCornerShape(40.dp)
            ) {
                Text(
                    if (!showAnswer) "Show Answer" else "Hide Answer",
                    fontSize = 20.sp,
                    fontWeight = FontWeight.SemiBold,
                    color = Color.White
                )
            }


            Spacer(modifier = Modifier.height(24.dp))
        }
    }
}

File: non_native/lib/views/practice_screen.dart

import 'package:flutter/material.dart';
import 'package:provider/provider.dart';
import '../view_models/main_view_model.dart';
import '../models/flashcard.dart';

class PracticeScreen extends StatefulWidget {
  const PracticeScreen({super.key});

  @override
  State<PracticeScreen> createState() => _PracticeScreenState();
}

class _PracticeScreenState extends State<PracticeScreen> {
  int _currentIndex = 0;
  bool _showAnswer = false;

  void _nextCard(int total) {
    if (_currentIndex < total - 1) {
      setState(() {
        _currentIndex++;
        _showAnswer = false;
      });
    }
  }

  void _previousCard() {
    if (_currentIndex > 0) {
      setState(() {
        _currentIndex--;
        _showAnswer = false;
      });
    }
  }

  @override
  Widget build(BuildContext context) {
    final viewModel = context.watch<MainViewModel>();
    final flashcards = viewModel.flashcards;
    final total = flashcards.length;

    if (total == 0) {
      return Scaffold(
        backgroundColor: const Color(0xFFF0F0F0),
        body: Center(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              const Text("No flashcards available for this category."),
              const SizedBox(height: 20),
              ElevatedButton(
                onPressed: () => Navigator.of(context).pop(),
                child: const Text('Go Back'),
              ),
            ],
          ),
        ),
      );
    }

    final currentFlashcard = flashcards[_currentIndex];
    final text = _showAnswer ? currentFlashcard.answer : currentFlashcard.question;

    
    const baseSize = 45.0;
    const minSize = 14.0;
    final scale = (baseSize - (text.length * 0.25)).clamp(minSize, baseSize);
    final dynamicFontSize = scale;

    Color diffBgColor;
    Color diffTextColor;
    switch (currentFlashcard.difficulty.name) {
      case 'easy':
        diffBgColor = const Color(0x334CAF50); 
        diffTextColor = const Color(0xFF4CAF50); 
        break;
      case 'medium':
        diffBgColor = const Color(0x33FFA726); 
        diffTextColor = const Color(0xFFFFA726); 
        break;
      case 'hard':
        diffBgColor = const Color(0x33F44336); 
        diffTextColor = const Color(0xFFF44336); 
        break;
      default:
        diffBgColor = Colors.grey.withOpacity(0.3);
        diffTextColor = Colors.grey;
    }

    return Scaffold(
      backgroundColor: const Color(0xFFF0F0F0),
      appBar: AppBar(
        title: const Text("Practice"),
        backgroundColor: const Color(0xFFF0F0F0),
        centerTitle: true,
        leading: IconButton(
          icon: const Icon(Icons.close, color: Colors.black),
          onPressed: () => Navigator.of(context).pop(),
        ),
        bottom: PreferredSize(
          preferredSize: const Size.fromHeight(1.0),
          child: Divider(color: Colors.black.withOpacity(0.07), thickness: 1.0),
        ),
      ),
      body: Padding(
        padding: const EdgeInsets.symmetric(horizontal: 24.0),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.center,
          children: [
            const SizedBox(height: 40.0),
            
            Text('${_currentIndex + 1}/$total', style: const TextStyle(fontWeight: FontWeight.w500, color: Colors.grey, fontSize: 16.0)),
            const SizedBox(height: 20.0),
            
            LinearProgressIndicator(
              value: (_currentIndex + 1) / total.toDouble(),
              backgroundColor: Colors.grey[300],
              valueColor: const AlwaysStoppedAnimation<Color>(Color(0xFF1280ED)),
              minHeight: 8.0,
              borderRadius: BorderRadius.circular(4.0),
            ),
            const SizedBox(height: 60.0),
            
            Expanded(
              flex: 5,
              child: GestureDetector(
                onTap: () {
                  setState(() {
                    _showAnswer = !_showAnswer;
                  });
                },
                child: Container(
                  width: double.infinity,
                  decoration: BoxDecoration(
                    color: Colors.white,
                    borderRadius: BorderRadius.circular(20.0),
                    boxShadow: [
                      BoxShadow(
                        color: Colors.black.withOpacity(0.1),
                        blurRadius: 10,
                        offset: const Offset(0, 5),
                      ),
                    ],
                  ),
                  padding: const EdgeInsets.all(20.0),
                  alignment: Alignment.topLeft,
                  child: Column(
                    crossAxisAlignment: CrossAxisAlignment.start,
                    children: [
                      
                      Container(
                        decoration: BoxDecoration(
                          color: diffBgColor,
                          borderRadius: BorderRadius.circular(8.0),
                        ),
                        padding: const EdgeInsets.symmetric(horizontal: 12.0, vertical: 6.0),
                        child: Text(
                          
                          '${currentFlashcard.difficulty.name[0].toUpperCase()}${currentFlashcard.difficulty.name.substring(1)}',
                          style: TextStyle(color: diffTextColor, fontWeight: FontWeight.bold),
                        ),
                      ),
                      const SizedBox(height: 24.0),
                      
                      Text(
                        text,
                        style: TextStyle(
                          fontSize: dynamicFontSize,
                          fontWeight: FontWeight.w500,
                          color: Colors.black,
                        ),
                        textAlign: TextAlign.start,
                      ),
                    ],
                  ),
                ),
              ),
            ),
            const SizedBox(height: 36.0),
            
            SizedBox(
              width: MediaQuery.of(context).size.width * 0.75,
              height: 70.0,
              child: ElevatedButton(
                onPressed: () {
                  setState(() {
                    _showAnswer = !_showAnswer;
                  });
                },
                style: ElevatedButton.styleFrom(
                  backgroundColor: _showAnswer ? const Color(0xFFFFA726) : const Color(0xFF1280ED),
                  shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(40.0)),
                ),
                child: Text(
                  _showAnswer ? "Hide Answer" : "Show Answer",
                  style: const TextStyle(fontSize: 20.0, fontWeight: FontWeight.w600, color: Colors.white),
                ),
              ),
            ),
            const SizedBox(height: 24.0),
            
            Padding(
              padding: const EdgeInsets.only(bottom: 12.0),
              child: Row(
                mainAxisAlignment: MainAxisAlignment.spaceBetween,
                children: [
                  SizedBox(
                    width: 170.0,
                    height: 50.0,
                    child: ElevatedButton(
                      onPressed: _currentIndex > 0 ? _previousCard : null,
                      style: ElevatedButton.styleFrom(
                        backgroundColor: const Color(0xFF1280ED),
                        shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(12.0)),
                      ),
                      child: Row(
                        mainAxisAlignment: MainAxisAlignment.center,
                        children: [
                          const Icon(Icons.arrow_back, color: Colors.white),
                          const SizedBox(width: 8.0),
                          Text("Previous", style: TextStyle(color: Colors.white.withOpacity(_currentIndex > 0 ? 1.0 : 0.5), fontSize: 16.0)),
                        ],
                      ),
                    ),
                  ),
                  SizedBox(
                    width: 170.0,
                    height: 50.0,
                    child: ElevatedButton(
                      onPressed: _currentIndex < total - 1 ? () => _nextCard(total) : null,
                      style: ElevatedButton.styleFrom(
                        backgroundColor: const Color(0xFFFFA726),
                        shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(12.0)),
                      ),
                      child: Row(
                        mainAxisAlignment: MainAxisAlignment.center,
                        children: [
                          Text("Next", style: TextStyle(color: Colors.white.withOpacity(_currentIndex < total - 1 ? 1.0 : 0.5), fontSize: 16.0)),
                          const SizedBox(width: 8.0),
                          const Icon(Icons.arrow_forward, color: Colors.white),
                        ],
                      ),
                    ),
                  ),
                ],
              ),
            ),
          ],
        ),
      ),
    );
  }
}